core/metric.go

```diff
@@ -0,0 +1,197 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	api "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk/metric"
+)
+
+const (
+	meterName     = "github.com/hyperledger-labs/yui-relayer"
+	namespaceRoot = "relayer"
+	fallbackAddr  = "localhost:0"
+)
+
+var (
+	meterProvider *metric.MeterProvider
+	meter         api.Meter
+
+	metricSrcProcessedBlockHeight *atomic.Int64
+	metricDstProcessedBlockHeight *atomic.Int64
+	metricSrcBacklog              PacketInfoList // take care. this is not thread safe.
+	metricDstBacklog              PacketInfoList // take care. this is not thread safe.
+	metricSrcBacklogSize          *atomic.Int64
+	metricDstBacklogSize          *atomic.Int64
+	metricSrcBacklogOldestHeight  *atomic.Int64
+	metricDstBacklogOldestHeight  *atomic.Int64
+
+	instProcessedBlockHeight    api.Int64ObservableGauge
+	instBacklogSize             api.Int64ObservableGauge
+	instBacklogOldestHeight     api.Int64ObservableGauge
+	instReceivePacketsFinalized api.Int64Counter
+)
+
+func InitializeMetrics(addr string) error {
+	if addr == "" {
+		addr = fallbackAddr
+	}
+
+	var err error
+
+	if meterProvider, err = NewPrometheusMeterProvider(addr); err != nil {
+		return err
+	}
+
+	meter = meterProvider.Meter(meterName)
+
+	metricSrcProcessedBlockHeight = &atomic.Int64{}
+	metricDstProcessedBlockHeight = &atomic.Int64{}
+	metricSrcBacklogSize = &atomic.Int64{}
+	metricDstBacklogSize = &atomic.Int64{}
+	metricSrcBacklogOldestHeight = &atomic.Int64{}
+	metricDstBacklogOldestHeight = &atomic.Int64{}
+
+	// create the instrument "relayer.processed_block_height"
+	name := fmt.Sprintf("%s.processed_block_height", namespaceRoot)
+	callback := func(ctx context.Context, observer api.Int64Observer) error {
```

Comment:
Thank you for a good suggestion! However there is a reason why I can't adopt this soon.

An instance of `SyncHeaders` is created, used, and dropped for each relay cycle.
Therefore it is meaningless to observer `SyncHeaders` to obtain metrics.
Not only `SyncHeaders` but all other states are also short-lived and not held longer than one relay cycle.
Therefore we have to add some "global variables" to store metrics.

Reply:
I forgot to say that we basically want to keep the core structure of the relayer stateless.

Reply:
I noticed my big mistake ... when I tried to move `metric.go` to another directory.
`metric.go` can be referred to by any other components in the relayer.
Therefore, to avoid cyclic reference, `metric.go` itself should not depends on any types or functions in the relayer.

Reply:
Thanks for explaining.

I have a question and a comment.

#### Question

Are you saying that `SyncHeader` is not **ALWAYS** long-lived?

`SyncHeader` seems long-lived structure when a relayer starts in service mode.

https://github.com/siburu/yui-relayer/blob/e6c5f9eca4fbde03c04218149776f068e948bb1d/core/service.go#L13
https://github.com/siburu/yui-relayer/blob/e6c5f9eca4fbde03c04218149776f068e948bb1d/core/naive-strategy.go#L65

It's OK that metrics are availbable only running in service mode.

#### Comment

You can avoid cyclic reference by defining an interface like `type SubjectI interface { Observe() }` at metrics.go. (like [here](https://github.com/hyperledger-labs/yui-relayer/pull/98#discussion_r1301876499))
Other types depend on `SubjectI`. metrics.go will keep independence from other go files/packages.

Reply:
@mizuochik 
Right. I was wrong about the lifespan of `SyncHeader`.

Please forget the 3 comments that start from https://github.com/hyperledger-labs/yui-relayer/pull/98#discussion_r1305066551.
My current position is https://github.com/hyperledger-labs/yui-relayer/pull/98#discussion_r1305350013.

I think the gauges that I added are more suitable for synchronous gauge, but the current release of OTel-go doesn't support synchronous gauge yet.
Therefore I defined our own implementation of synchronous gauge in `metrics/sync_gauge.go`.

I believe synchronous gauge will be supported by OTel-go in the near future.
https://github.com/open-telemetry/opentelemetry-specification/pull/3540

Reply:
Thanks. I agreed! 

`SyncGauge` seems suitable to me as well.